File: controllers/turnstile.go
package controllers

import (
	"encoding/json"
	"errors"
	"net/http"
	"net/url"
	"os"
	"strings"

	"github.com/beego/beego/v2/server/web"
)

const turnstileVerifyURL = "https://challenges.cloudflare.com/turnstile/v0/siteverify"

type turnstileVerifyResponse struct {
	Success    bool     `json:"success"`
	ErrorCodes []string `json:"error-codes"`
}

func turnstileSecret() string {
	return strings.TrimSpace(firstNonEmpty(
		os.Getenv("TURNSTILE_SECRET_KEY"),
		web.AppConfig.DefaultString("turnstile_secret_key", ""),
	))
}

func verifyTurnstileToken(token, remoteIP string) error {
	secret := turnstileSecret()
	if secret == "" {
		return errors.New("turnstile is not configured")
	}
	if strings.TrimSpace(token) == "" {
		return errors.New("turnstile token is required")
	}

	form := url.Values{}
	form.Set("secret", secret)
	form.Set("response", strings.TrimSpace(token))
	if strings.TrimSpace(remoteIP) != "" {
		form.Set("remoteip", strings.TrimSpace(remoteIP))
	}

	resp, err := http.PostForm(turnstileVerifyURL, form)
	if err != nil {
		return errors.New("failed to verify turnstile token")
	}
	defer resp.Body.Close()

	var verifyResp turnstileVerifyResponse
	if err := json.NewDecoder(resp.Body).Decode(&verifyResp); err != nil {
		return errors.New("invalid turnstile verification response")
	}
	if !verifyResp.Success {
		return errors.New("turnstile verification failed")
	}

	return nil
}

File: packages/api/src/helpers/turnstile.ts
const TURNSTILE_VERIFY_URL = "https://challenges.cloudflare.com/turnstile/v0/siteverify";

export async function verifyTurnstileToken(token: string, remoteIp?: string): Promise<boolean> {
  const secret = process.env.TURNSTILE_SECRET_KEY;
  if (!secret) throw new Error("TURNSTILE_SECRET_KEY is not configured");
  if (!token) return false;

  const form = new URLSearchParams({
    secret,
    response: token,
  });
  if (remoteIp) form.set("remoteip", remoteIp);

  try {
    const res = await fetch(TURNSTILE_VERIFY_URL, {
      method: "POST",
      body: form,
      headers: { "Content-Type": "application/x-www-form-urlencoded" },
    });
    const data = await res.json() as { success?: boolean };
    return data.success === true;
  } catch {
    return false;
  }
}
